### ISSUE TITLE
Problematic uses of `$(MAKE)`

### ISSUE THREAD

**Issue by jaredcdavis** [acl2/acl2#531]
I think there are problems with certain uses of `$(MAKE)` in `books/GNUmakefile`.

Consider in particular the `manual` target:

```
manual:
	$(MAKE) USE_QUICKLISP=1 doc/top.cert
```

Now consider what happens to the hapless user who types something like this:

```
$ make clean
$ make -j 100 basic centaur manual
```

At this point:

   - `make` starts up 99 threads to certify the books that `basic` and `centaur` depend on.
   - With the other thread, it starts to try to build the `manual` target, which has no dependencies.  But to do this, it **invokes a second make in the background** and instructs it to build `doc/top`.

Chaos ensues.  The second `make` is trying to build a bunch of the same books that the first `make` is building (i.e., the books that support `basic`, `centaur`, etc.)  As both `make`s are running, at any given point in time, just about the same books are ready to be certified.  So the two `make` commands frequently try to build the exact same book.

This results in mysterious certification failures as shown below.

So I think we need to be very careful about using `$(MAKE)`.  It seems likely to be OK in cases such as:

```
nonstd/workshops/1999/calculus/book/tree.cert
```

Where we're just using it to build a single file in a subdirectory, as long as we've worked to make sure that we have the dependencies of `tree.cert` known to the top level `GNUmakefile`.  Similarly, I think the `milawa` targets are also OK because we explicitly exclude the Milawa books and the Makefile knows what the dependencies are.

But the following seem very suspicious and may have the problem described above:

```
quicklisp:
	$(MAKE) REALLY_USE_QUICKLISP=1 $(QUICKLISP_DIR)/top.cert

everything:
	$(MAKE) all $(ADDED_BOOKS) $(ADDED_BOOK_TARGETS)

manual:
	$(MAKE) USE_QUICKLISP=1 doc/top.cert
```

I'm less sure about things like `no_hons_error` and `no_ansi_error`, but these seem like really strange targets in the first place; why not just use `$(error ...)` here instead?

Example failures, as mentioned above:

```
   for /share/apps/fv/jared/xd/acl2/books/defexec/other-apps/records/records.lisp

   | ACL2 Error in ( INCLUDE-BOOK "misc/total-order" ...):  There is no
   | certificate on file for 
   | "/share/apps/fv/jared/xd/acl2/books/misc/total-order.lisp".  This is
   | illegal because we are currently attempting certify-book; see :DOC
   | certify-book.

   for /share/apps/fv/jared/xd/acl2/books/std/util/look-up.lisp   

   | ACL2 Error in ( INCLUDE-BOOK "support" ...):  There is no certificate
   | on file for "/share/apps/fv/jared/xd/acl2/books/std/util/support.lisp".
   | This is illegal because we are currently attempting certify-book; see
   | :DOC certify-book.

   for /share/apps/fv/jared/xd/acl2/books/misc/expander.lisp

   |  (2B58500187C0) : 0 (PRINT-CALL-HISTORY :CONTEXT NIL :PROCESS NIL :ORIGIN NIL :DETAILED-P NIL :COUNT 10000 :START-FRAME-NUMBER 0 :STREAM #<SYNONYM-STREAM to *TERMINAL-IO* #x30200351B5DD> :PRINT-LEVEL 2 :PRINT-LENGTH 5 :SHOW-INTERNAL-FRAMES NIL :FORMAT :TRADITIONAL) 853
   |  (2B5850018928) : 1 (CALL-CHECK-REGS CCL:PRINT-CALL-HISTORY :DETAILED-P NIL :COUNT 10000) 221
   |  (2B5850018960) : 2 (CHEAP-EVAL (CCL:PRINT-CALL-HISTORY :DETAILED-P NIL :COUNT *CCL-PRINT-CALL-HISTORY-COUNT*)) 101
   |  (2B5850018998) : 3 (PRINT-CALL-HISTORY) 133
   |  (2B58500189A8) : 4 (OUR-ABORT #<CCL::SIMPLE-FILE-ERROR #x302003CD0E4D> OUR-ABORT) 173
   |  (2B58500189E8) : 5 (BREAK-LOOP-HANDLE-ERROR #<CCL::SIMPLE-FILE-ERROR #x302003CD0E4D> 5957287424344) 637
   |  (2B5850018A98) : 6 (%ERROR #<CCL::SIMPLE-FILE-ERROR #x302003CD0E4D> NIL 5957287424344) 357
   |  (2B5850018AC0) : 7 (DELETE-EXPANSION-FILE "/share/apps/fv/jared/xd/acl2/books/misc/expander@expansion.lsp" ACL2_INVISIBLE::|The Live State Itself|) 53
   |  (2B5850018B00) : 8 (CERTIFY-BOOK-FN "expander" ? T NIL :DEFAULT NIL NIL NIL NIL :DEFAULT :DEFAULT ACL2_INVISIBLE::|The Live State Itself|) 44277
   | ...
   | ***********************************************
   | ************ ABORTING from raw Lisp ***********
   | Error:  No such file or directory : "/share/apps/fv/jared/xd/acl2/books/misc/expander@expansion.lsp"
   | NOTE: See above for backtrace.
   | ***********************************************

   for some other file:

   | ACL2 Error in ACL2-INTERFACE:  Evaluation aborted.  To debug see :DOC
   | HARD ACL2 ERROR in INCLUDE-RAW:  Load of "base-raw.lsp" failed with
   | print-gv, see :DOC trace, and see :DOC wet.
   | the following message:
   | 
   | File #P"/share/apps/fv/jared/xd/acl2/books/centaur/quicklisp/asdf-
   | home/cache/common-lisp/ccl-1.11-f96-linux-x64/share/apps/fv/jared/xd/acl2/book\
   | s/centaur/quicklisp/bundle/software/flexi-streams-1.0.15/external-
   | format-TMP.lx64fsl" does not exist.
```

**Comment by MattKaufmann** [comment]
Good catch.  A simple solution is to comment out the quicklisp and manual targets (the comments then would show what to do) and to add a comment for the everything target that it should only be used by itself if -j is given a value greater than 1.

Shall I make these changes?

**Comment by jaredcdavis** [comment]
I'm fixing...


On Thu, Oct 22, 2015 at 10:02 AM, MattKaufmann <notifications@github.com>
wrote:

> Good catch. A simple solution is to comment out the quicklisp and manual
> targets (the comments then would show what to do) and to add a comment for
> the everything target that it should only be used by itself if -j is given
> a value greater than 1.
>
> Shall I make these changes?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/acl2/acl2/issues/531#issuecomment-150254361>.
>



-- 
Jared C. Davis <jared@cs.utexas.edu>
11410 Windermere Meadows
Austin, TX 78759
http://www.cs.utexas.edu/users/jared/

**Comment by jaredcdavis** [comment]
I think this is fixed as of ae4a50cdda0e3a260febea0c2f6545f7c4c57e6e
